File: iosApp/iosApp/ProfileViewModel.swift
import SwiftUI
import shared

class ProfileViewModel: ObservableObject {
    
    // Estados
    @Published var authState: AuthState = .idle
    @Published var searchState: SearchState = .idle
    @Published var projectsState: ProjectsState = .idle
    
    enum AuthState : Equatable {
        case idle
        case loading
        case success(authData: AuthData)
        case error(message: String)
    }
    
    enum SearchState: Equatable  {
        case idle
        case loading
        case success(user: SelectedUserProfile)
        case error(message: String)
    }
    
    enum ProjectsState: Equatable  {
        case idle
        case loading
        case success(projects: [Project])
        case error(message: String)
    }

    // Función para manejar el callback de OAuth
    func handleAuthCallback(code: String) {
        DispatchQueue.global().async {
            do {
                print("[VIEWMODEL] Code: \(code)")
                self.updateAuthState(.loading)
                
                let authData = try Api42().handleCallbackWrapper(code: code)
                
                DispatchQueue.main.async {
                    // Acceso CORRECTO a las propiedades (snake_case igual que en Kotlin)
                    SessionManager.shared.access_token = authData.access_token
                    SessionManager.shared.refresh_token = authData.refresh_token
                    SessionManager.shared.user_id = authData.userId as? KotlinInt // userId es Int en Kotlin
                    SessionManager.shared.user_login = authData.userLogin
                    SessionManager.shared.user_image_url = authData.imageUrl
                                    
                    self.updateAuthState(.success(authData: authData))
                                    
                    let userId = SessionManager.shared.user_id?.int32Value ?? -1
                    print("[VIEWMODEL]: Logged as: \(SessionManager.shared.user_login ?? ""), id: \(userId)")
                }
                
            } catch {
                print("[VIEWMODEL] Auth error: \(error)")
                self.updateAuthState(.error(message: error.localizedDescription))
                SessionManager.shared.clearSession()
            }
        }
    }
        
    func searchForUser(login: String) {
        DispatchQueue.global().async {
            do {
                self.updateSearchState(.loading)
                print("[VIEWMODEL] Searching for login: \(login)")
                
                let user = try Api42().searchUserWrapper(login: login)
                SessionManager.shared.selectedUserProfile = user
                
                DispatchQueue.main.async {
                    self.updateSearchState(.success(user: user))
                    print("[VIEWMODEL] User found: \(user.login)")
                }
            } catch let error as Api42.UserNotFoundException {
                self.updateSearchState(.error(message: "Usuario no encontrado"))
            } catch {
                self.updateSearchState(.error(message: error.localizedDescription))
            }
        }
    }

    func loadProjectsForUser(login: String) {
        DispatchQueue.global().async {
            do {
                self.updateProjectsState(.loading)
                print("[VIEWMODEL] Loading projects for: \(login)")
                
                let projects = try Api42().getProjectsWrapper(login: login)
                
                DispatchQueue.main.async {
                    if var selectedUser = SessionManager.shared.selectedUserProfile {
                        selectedUser.projects = projects
                        SessionManager.shared.selectedUserProfile = selectedUser
                    }
                    self.updateProjectsState(.success(projects: projects))
                    print("[VIEWMODEL] Projects loaded: \(projects.count)")
                }
            } catch {
                print("[VIEWMODEL] Projects error: \(error)")
                self.updateProjectsState(.error(message: error.localizedDescription))
                
                if error.localizedDescription.contains("401") {
                    self.refreshTokenAndRetry(login: login)
                }
            }
        }
    }
        
    private func refreshTokenAndRetry(login: String) {
        Task {
            do {
                let success = try await Api42().refreshTokenWrapper()
                if success as! Bool {
                    DispatchQueue.main.async {
                        self.loadProjectsForUser(login: login)
                    }
                }
            } catch {
                print("[VIEWMODEL] Token refresh failed: \(error)")
            }
        }
    }
    
    // MARK: - State Management
        
    private func updateAuthState(_ state: AuthState) {
        DispatchQueue.main.async {
            self.authState = state
        }
    }
    
    private func updateSearchState(_ state: SearchState) {
        DispatchQueue.main.async {
            self.searchState = state
        }
    }
    
    private func updateProjectsState(_ state: ProjectsState) {
        DispatchQueue.main.async {
            self.projectsState = state
        }
    }
    
    func resetState() {
        DispatchQueue.main.async {
            self.authState = .idle
            self.searchState = .idle
            self.projectsState = .idle
        }
    }
    
    func clearSearch() {
        DispatchQueue.main.async {
            SessionManager.shared.selectedUserProfile = nil
            self.searchState = .idle
        }
    }
    
    func clearAuthError() {
        DispatchQueue.main.async {
            if case .error = self.authState {
                self.authState = .idle
            }
        }
    }
}

File: androidApp/src/main/java/com/example/intrapp/android/ProfileViewModel.kt
package com.example.intrapp.android


import android.util.Log
import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.example.intrapp.Api42
import com.example.intrapp.AuthData
import com.example.intrapp.SelectedUserProfile
import com.example.intrapp.SessionManager
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.launch


class ProfileViewModel : ViewModel() {

    // Estados
    private val _authState = MutableStateFlow<AuthState>(AuthState.Idle)
    val authState: StateFlow<AuthState> = _authState

    private val _searchState = MutableStateFlow<SearchState>(SearchState.Idle)
    val searchState: StateFlow<SearchState> = _searchState

    sealed class AuthState {
        object Idle : AuthState()
        object Loading : AuthState()
        data class Success(val authData: AuthData) : AuthState()
        data class Error(val message: String) : AuthState()
    }

    sealed class SearchState {
        object Idle : SearchState()
        object Loading : SearchState()
        data class Success(val user: SelectedUserProfile) : SearchState()
        data class Error(val message: String) : SearchState()
    }

    fun handleAuthCallback(code: String) {
        viewModelScope.launch {
            _authState.value = AuthState.Loading
            try {
                println("[VIEWMODEL] Code: $code")

                // Llamar a API42 para obtener tokens y datos básicos
                val authData = Api42().handleCallback(code)

                // Guardar datos en SessionManager
                SessionManager.apply {
                    access_token = authData.access_token
                    refresh_token = authData.refresh_token
                    user_id = authData.userId
                    user_login = authData.userLogin
                    user_image_url = authData.imageUrl
                }
                // Indicar que la autenticación fue exitosa
                _authState.value = AuthState.Success(authData)

                println("[VIEWMODEL]:handlecallback() =  Authdata for : ${SessionManager.user_login}, id: ${SessionManager.user_id}")
                println("[VM]handlecallback() = Image link: ${SessionManager.user_image_url}")
                println("[VM]handlecallback() = Access Token: ${SessionManager.access_token}")
                println("[VM]handlecallback() = Refresh Token: ${SessionManager.refresh_token}")

            } catch (e: Exception) {
                Log.e("ProfileViewModel", "Error en autenticación", e)
                _authState.value = AuthState.Error(e.message ?: "Error desconocido")
            }
        }
    }


    // Función para buscar usuarios
    fun searchForUser(login: String) {
        viewModelScope.launch {
            _searchState.value = SearchState.Loading
            try {
                val user = Api42().searchUser(login)
                SessionManager.selectedUserProfile = user
                _searchState.value = SearchState.Success(user)
            } catch (e: Exception) {
                _searchState.value = SearchState.Error(e.message ?: "Error buscando usuario")
            }
        }
    }

    fun resetState() {
        _authState.value = AuthState.Idle
        _searchState.value = SearchState.Idle
    }


    // Clase de excepción personalizada
    class UserNotFoundException : Exception("Usuario no encontrado")

    // Funcion para cargar proyectos de un usuario
    suspend fun loadProjectsForUser(login: String) {
        try {
            val projects = Api42().getProjectsForUser(login)
            SessionManager.selectedUserProfile = SessionManager.selectedUserProfile?.copy(
                projects = projects
            )
        } catch (e: Exception) {
            Log.e("ProfileViewModel", "Error loading projects for user", e)
            throw e
        }
    }

    // Función para limpiar errores anteriores
    fun clearAuthError() {
        SessionManager.lastAuthError = null
    }

    // Función para limpiar los resultados de búsqueda
    fun clearSearch() {
        SessionManager.selectedUserProfile = null
        _searchState.value = SearchState.Idle
    }


}


/*
viewModelScope.launch: Ejecuta el código en un coroutine ligado al ciclo de vida del ViewModel.
 */